### PR TITLE
avoid tombstoned rows

### PR DESCRIPTION
**Pull request by nziebart**
**Goals (and why)**:
Addresses part of https://github.com/palantir/atlasdb/issues/2691 by paging over tombstones, rather than making Cassandra do an unbounded range scan.

**Implementation Description (bullets)**:
- Impose an upper bound on the CQL range scan used for sweep.
- Use thrift to determine a safe upper bound (see comments in `GetCellTimestamps` for an explanation of why this works.

**Concerns (what feedback would you like?)**:

**Where should we start reviewing?**:

**Priority (whenever / two weeks / yesterday)**:

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/2732)

### CODE REVIEW

File: atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/OneNodeDownGetTest.java [hunk 1]
```diff
@@ -112,14 +112,14 @@ public void getRangeOfTimestampsThrows() {
...
+                .hasMessageContaining("requires all Cassandra nodes to be up and available.");
     }
 
     @Test
     public void getAllTimestampsThrows() {
         assertThatThrownBy(() -> OneNodeDownTestSuite.kvs.getAllTimestamps(OneNodeDownTestSuite.TEST_TABLE,
                 ImmutableSet.of(OneNodeDownTestSuite.CELL_1_1), Long.MAX_VALUE))
                 .isExactlyInstanceOf(InsufficientConsistencyException.class)
-                .hasMessage("This get operation requires ALL Cassandra nodes to be up and available.");
+                .hasMessageContaining("requires ALL Cassandra nodes to be up and available.");
```
Comment (gsheasby) [inline atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/OneNodeDownGetTest.java]: Our insufficient consistency exception messages are not sufficiently consistent! ("all" vs "ALL")

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/sweep/GetCellTimestamps.java [hunk 2]
```diff
@@ -63,14 +76,51 @@ public GetCellTimestamps(
...
+     * return the tombstoned rows (they will just be empty KeySlices), which allows us to page over them. So, we first
+     * execute a thrift range scan to determine an upper bound on the CQL range scan, and page over the data until we
+     * find some live cells.
+     * <p>
+     * This is a hack (and a perf hit) until we have a better solution to avoid scanning massive numbers of
+     * tombstones (ideally this will involve catching some exception and reducing the range appropriately).
+     */
+    private void fetchBatchOfTimestampsBeginningAtStartRow() {
+        Optional<byte[]> rangeEnd;
+        for (byte[] rangeStart = startRowInclusive; timestamps.isEmpty(); rangeStart = rangeEnd.get()) {
```
Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/sweep/GetCellTimestamps.java]: This is slightly confusing. I wonder if the following would read better:
```
byte[] rangeStart = startRowInclusive;
while (timestamps.isEmpty()) {
    Optional<byte> rangeEnd = determineSafeRangeEnd(rangeStart);
    // as in lines 101-106
    rangeStart = rangeEnd.get();
}
```

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/sweep/GetCellTimestamps.java [hunk 3]
```diff
@@ -63,14 +76,51 @@ public GetCellTimestamps(
...
+
+            List<CellWithTimestamp> batch = cqlExecutor.getTimestamps(tableRef, rangeStart, rangeEnd.get(), batchHint);
+            timestamps.addAll(batch);
+        }
+    }
 
-        timestamps.addAll(batch);
+    private Optional<byte[]> determineSafeRangeEnd(byte[] rangeStart) {
+        KeyRange keyRange = new KeyRange().setStart_key(rangeStart).setEnd_key(new byte[0]).setCount(batchHint);
+        SlicePredicate slicePredicate = SlicePredicates.create(SlicePredicates.Range.ALL, SlicePredicates.Limit.ONE);
```
Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/sweep/GetCellTimestamps.java]: Does the `Limit.ONE` effectively mean that the `List<KeySlice>` will just contain up to `batchHint` row keys with 0-1 columns?

Comment (gsheasby) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/sweep/GetCellTimestamps.java]: Do we need to be concerned about loading values with this query?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/sweep/GetCellTimestamps.java [hunk 4]
```diff
@@ -63,14 +76,51 @@ public GetCellTimestamps(
...
+                return;
+            }
+
+            List<CellWithTimestamp> batch = cqlExecutor.getTimestamps(tableRef, rangeStart, rangeEnd.get(), batchHint);
+            timestamps.addAll(batch);
+        }
+    }
 
-        timestamps.addAll(batch);
+    private Optional<byte[]> determineSafeRangeEnd(byte[] rangeStart) {
```
Comment (tpetracca) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/sweep/GetCellTimestamps.java]: Can we comment describing desired logic? As far as I can tell it's:
> Ask for `batchHint` many row keys after `rangeStart`, and return the final key. If there are no keys return `Optional.empty()`. Keys for rows that entirely consist of tombstones are counted and one will potentially be returned.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/sweep/GetCellTimestamps.java (hunk 3 above)
Comment (nziebart) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/sweep/GetCellTimestamps.java]: Yes, we need to be concerned with loading values here. I wonder if it's possible to execute a range slice query with a limit of zero..

**Pull request merged**